Group Solution to assignment #5

VolhaAnishchanka opened the pull request:
Some comments on the code still need to be added

VolhaAnishchanka committed:
Add files via upload

Some comments on the code still need to be added

YakimS commented on group/assignment5/AlphabetDiscoverer.java:
```diff
@@ -0,0 +1,86 @@
+    private void addAllCharsInDictionaryToGraph(List<String> dictionary) {
+        for (String s : dictionary) {
+            for (int i = 0 ; i < s.length() ; i++) {
+                Character vertex = s.charAt(i);
```
The character should be called "currentChar" or similar name, but not "vertex".
This is because this Character doesn't represent a vertex, but a value which a vertex will hold.

narciuch replied:
Agreed. 
Perhaps also rename 's' to 'currentWord', to make it clearer what the entity is?

narciuch replied:
Done

YakimS commented on group/assignment5/AlphabetDiscoverer.java:
```diff
@@ -0,0 +1,86 @@
+        for (String s : dictionary) {
+            for (int i = 0 ; i < s.length() ; i++) {
+                Character vertex = s.charAt(i);
+                if (!graph.containsVertex(vertex)) {
```
I think we should move this check to the function "addVertex" in Graph class, and comment in the class comment that says: "Graph vertices are unique - only one vertex can be created for each T value".

narciuch replied:
Done by changing in and out vertices into sets

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+     * @return topological ordered list of vertecies of a graph
+     */
+    public List<T> getTopologicalOrder() {
+        List<T> alphabet = new ArrayList<>();
```
Shouldn't be called "alphabet" because the List<T> can represent many things and not just alphabets.
A proper name, in my opinion, is "verticesValues".
The use of the word "alphabet" should be changed in all the functions in Graph.class

YakimS replied:
Done

YakimS commented on group/assignment5/Constraint.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public String toString() {
+        return "Character " + second + " cannot be ordered before " + first;
```
Replace "Character" with "Value".  Because T can be a different type.

Copy instead of row 18:
return "Value " + second + " cannot be ordered before " + first;

narciuch replied:
Done

YakimS commented on group/assignment5/Constraint.java:
```diff
@@ -0,0 +1,38 @@
+ * @param <T>
+ */
+public class Constraint<T extends Comparable<T>> {
+    public T first;
```
Change maybe to "former" and "latter"?

narciuch replied:
Maybe 'from' and 'to' to make it even more explicit?

YakimS replied:
From-to is about direction, but this class is about order, so I think that former-latter or before-after is better.

YakimS replied:
Done

YakimS commented on group/assignment5/Constraint.java:
```diff
@@ -0,0 +1,38 @@
+package assignment5;
+
+/**
+ * This class represents a constraint made up of two vertecies of a graph.
```
This class is not about graphs vertices, but representing the relation between 2 Comparable values.

So maybe the class comment should be "This class represents an order-relation constraint between two Comparable objects from the same type"

narciuch replied:
Done

YakimS commented on group/assignment5/Constraint.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
We should use java conventions in equals(Object o), and hashCode() - use brackets in "if"s, ect.

YakimS replied:
Done

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Collections;
+
```
Add static import of "VertexState" to make the code more readable.

copy-paste this:
import static assignment5.VertexState.*;

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+    private Map<T, Vertex> vertecies;
+
+    public Graph() {
+        vertecies = new HashMap<>();
```
Misspelled - use "vertices" instead of "vertecies"

narciuch replied:
Done

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+ * easily.
+ */
+public class Graph<T extends Comparable<T>> {
+    private Map<T, Vertex> vertecies;
```
I'd call it "vertexByValue". Because the map is a lil bit more than just "vertices".

narciuch replied:
Done

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+        vertecies.put(value, new Vertex(value));
+    }
+
+    public Map<T, Vertex> getVertecies() {
```
This function never used. Can be deleted.

narciuch replied:
Done

YakimS commented on group/assignment5/VertexState.java:
```diff
@@ -0,0 +1,19 @@
+ * visited if there is no need to track the time when the search from this
+ * vertex has been started and finished which can be done using Visiting state
+ */
+public enum VertexState {
```
"DFSState", or "DFSVertexState" is a better name, I think.

YakimS replied:
Done

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+            return outdegrees;
+        }
+
+        public VertexState getState() {
```
"getDFSState()", maybe?

YakimS replied:
Done

YakimS commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+ * This class represents a directed graph data structure.
+ * A graph can be described as a set of vertecies. In case graph
+ * is directed each vertex has its indegree and outdegree vertecies.
+ * This class helps you to construct a directed graph and operate on it
```
I think this part is redundant:
"This class helps you to construct a directed graph and operate on it easily."

narciuch replied:
Done

narciuch commented on group/assignment5/AlphabetDiscoverer.java:
```diff
@@ -0,0 +1,86 @@
+        return minLength;
+    }
+
+    public List<Character> getAlphabet(List<String> dictionary) {
```
Might be good to add some documentation here, especially since this function is public. A description taken from the specification just to let readers know the expected pre and post conditions of the method. Maybe something like:

```
Given a dictionary (a list of words in lexicographic order) of all words in a possibly fictional language,
this method finds an alphabet (an ordered list of characters) of that language.

This language can contain any character (of the native char data type). 
Upper/lower case characters are treated differently. 
We assume standard lexicographical ordering (order by characters from left to right, if X is a prefix of Y then X is sorted before Y), but where the order of characters is unknown.
```

We could also mention how this function handles the cases of multiple consistent alphabets or no possible alphabets, referencing the getAllPossibleAlphabets function in relation to this?

narciuch commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+
+/**
+ * This class represents a directed graph data structure.
+ * A graph can be described as a set of vertecies. In case graph
```
I may be wrong, but I think 'vertecies' is misspelled. Isn't it 'vertices'?

narciuch commented on group/assignment5/Graph.java:
```diff
@@ -0,0 +1,218 @@
+     */
+    public class Vertex {
+        private T value;
+        private List<Vertex> indegrees;
```
Maybe it would be better to use a TreeSet to enforce a sorted collection as well as no duplicate values?

YakimS replied:
Right!

narciuch replied:
Done using Hashset - the collection in and out degrees don't need to be sorted

Commit:
Structured the project into src and test folders on the group submission branch

Commit:
Removed old copies of files

Commit:
Added some documentation and renamed some variables to clarify code

Commit:
Changed lists of in and out degrees into sets to ensure no duplicate neighbour vertices for a particular vertex

YakimS committed:
Add files via upload

YakimS committed:
Delete VertexState.java

YakimS commented on group/assignment5/src/VertexState.java:
```diff
@@ -1,17 +0,0 @@
- * visited if there is no need to track the time when the search from this
- * vertex has been started and finished which can be done using Visiting state
- */
-public enum VertexState {
```
I've moved this class into Graph.java

YakimS commented on group/assignment5/src/AlphabetDiscoverer.java:
```diff
@@ -17,9 +18,13 @@ private void buildGraphFromDictionary(List<String> dictionary) {
-        graph = new Graph<>();
-        addAllCharsInDictionaryToGraph(dictionary);
-        addConstraints(dictionary);
+        if(!dictionary.equals(currentDictionary))
```
Created this check so "buildGraphFromDictionary" won't be executed unnecessarily.

YakimS commented on group/assignment5/src/Graph.java:
```diff
@@ -1,5 +1,10 @@
-import java.util.*;
```
Please make sure not to use "*"

YakimS commented on group/assignment5/src/Graph.java:
```diff
@@ -45,103 +50,101 @@ public void addEdge(T parent, T child) {
-        dfsForAllTopologicalOrders(alphabet, alphabets);
-        vertexByValue.values().forEach(e -> e.setState(VertexState.UNVISITED));
-        return alphabets;
+        List<List<T>> allOrderedVerticesVal = new ArrayList<>();
```
I've moved each dfs function next to the caller function so the class will be more readable.

YakimS commented on group/assignment5/src/Constraint.java:
```diff
@@ -1,36 +1,41 @@
-public class Constraint<T extends Comparable<T>> {
-    public T from;
-    public T to;
+public class Constraint<T> {
```
I don't see a good reason T should extend Comparable, so I removed it.

YakimS replied:
If we leave it this way, we need to edit the class comment (remove the word "comparable").

VolhaAnishchanka committed:
Added test.

Added test for checking if constraints for circuits with common edge will be correct. In order to make sure that the same edge will not be added twice as a constraint.

stelaval reviewed (approved):
LGTM
